model/models.py

```diff
@@ -128,3 +130,87 @@ def forward(self, sub, rel):
 
 		score = torch.sigmoid(x)
 		return score
+
+	# Documentation
+
+
+class CompGCN_CTKGC(CompGCNBase):
+		'''
+		Implements the CTKGC Scoring Function as in the paper https://link.springer.com/article/10.1007/s10489-021-02438-8 and writes to the console the total Parameter uesd in the Network
+
+		The scoring Function works by
+			1. Build the Entity-relation matrix by multiplying sub_emb and rel_emb.transpose(1,0)
+			2. Then we perform a Convolution on Entity-relation matrix
+			3. We Aggregate all the obtained feature,
+			4. Convert the above feature maps to a 1D Vector
+			5. We project our 1D vector onto a candidate objects to obtain the predicted object embedding
+				1. We do 5. for all our entities (candidate objects)
+			5. Use the sigmoid function to calculate the score of the object embedding
+		'''
+		def __init__(self, edge_index, edge_type, params=None):
+			'''Init ComCGN Base and all the layers used in the Scoring function CTKGC'''
+			super(self.__class__, self).__init__(edge_index, edge_type, params.num_rel, params)
+
+			self.drop = torch.nn.Dropout(self.p.hid_drop)
+			self.hid_drop0 = torch.nn.Dropout(self.p.hid_drop)
+			self.fea_drop0 = torch.nn.Dropout2d(self.p.feat_drop)
+
+			filters = 32
+			kernelsize = (3, self.p.embed_dim)
+			hidden_size = filters *(self.p.embed_dim - kernelsize[0] +1) * (self.p.embed_dim - kernelsize[1] + 1)
+			self.conv2d0 = torch.nn.Conv2d(1, filters, kernelsize, 1, 0, bias=True)
+
+			self.bn0 = torch.nn.BatchNorm2d(1)
+			self.bn1 = torch.nn.BatchNorm2d(filters)
+			self.bn2 = torch.nn.BatchNorm1d(self.p.embed_dim)
+
+			self.fc = torch.nn.Linear(hidden_size, self.p.embed_dim)
+			total_params = sum(
+				param.numel() for param in self.parameters()
+			)
+			print("Total Parameter: " + str(total_params))
+
+
+		# Documentation
+		def forward(self, sub, rel):
+			'''
+
+			Parameters
+			----------
+			sub: idx of entities
+			rel: idx of relation
+
+			Returns
+			-------
+			a list of entities and the percentage of there being a relation of rel between sub and our entities
+
+			Inline comments describe the size of the tensor after manipulation with:
+				bs = batch_size
+				num_ent = Number of entities
+				embed_dim = embedding dimension
+			'''
+
+			sub_emb, rel_emb, all_ent = self.forward_base(sub, rel, self.drop,self.drop)
+			sub_emb = sub_emb.view(-1, self.p.embed_dim, 1)
+			rel_emb = rel_emb.view(-1, 1, self.p.embed_dim)
+
+			x = torch.bmm(sub_emb,rel_emb)
```

Comment:
I dont exactly see, where you transposed rel_emb or sub_emb here

Could you ellaborate on that?

Reply:
we are changing the view on the sub_emb this should do the same as transpose, because we only have a 1D Vector, and view should be faster

Reply:
That is alright then, performance improvements are always welcome. Maybe add a comment explaining that for better readability